### PR TITLE
add probes on fd_install to catch accept events before kretprobe was installed

### PR DESCRIPTION
**Pull request by alban**
First start a process that will block in the `accept()` syscall:
```
$ busybox nc -l -p 8181
```

Once the `accept()` syscall has started, start the tcptracer-bpf tool:
```
$ sudo ./tracer $(pidof busybox)
Ready
```

Makes the `accept()` syscall return:
```
$ busybox nc 127.0.0.1 8181
```

Notice that we get the "connect" event as normal. The "accept" event is missing because of #10 but we get the new event "fdinstall" that says that the process 16389 received the new file descriptor 4:
```
224557401988314 cpu#2 connect 16402 busybox 127.0.0.1:45488 127.0.0.1:8181 4026531969
224557402151953 cpu#3 fdinstall 16389 busybox 0.0.0.0:0 0.0.0.0:0 4
```

Weave Scope would need to iterate over `/proc` to find all the processes blocked on the `accept` syscall (`/proc/$pid/wchan` will show `inet_csk_accept`) and then request the tracker to monitor [fd_install](https://github.com/torvalds/linux/blob/v4.11/fs/file.c#L625-L630) on them:
```
  t.AddFdInstallWatcher(pid)
```
We don't notify [fd_install](https://github.com/torvalds/linux/blob/v4.11/fs/file.c#L625-L630) events for all processes because that would be too much. Instead, Scope would only request the notification for specific pids. Internally, this uses a hash map to keep track of the pids to monitor or not.

We don't need any new [offset guessing](https://github.com/weaveworks/tcptracer-bpf/blob/master/pkg/tracer/offsetguess.go) for this because we don't dereference any kernel structures in the fd_install kprobes.

This is for https://github.com/weaveworks/tcptracer-bpf/issues/10

Depends on:
- [x] https://github.com/iovisor/gobpf/pull/45
- [x] https://github.com/weaveworks/tcptracer-bpf/pull/40
- [x] https://github.com/weaveworks/build-tools/pull/97

TODO
- [x] test on Linux 4.4
- [x] add a field for `fd` in the event struct

/cc @iaguis @2opremio

### CODE REVIEW

**Comment by 2opremio** [comment]
> Weave Scope would need iterate over /proc to find all the processes blocked on the accept syscall (/proc/$pid/wchan will show inet_csk_accept) and then requests the tracker to monitor fd_install on them

Could we avoid this by making the probe on `fd_install` detect whether the file descriptor comes from a socket which was on LISTEN state?

**Comment by alban** [comment]
> Could we avoid this by making the probe on `fd_install` detect whether the file descriptor comes from a socket which was on LISTEN state?

Not sure how to do that, but this would involve more complicated offset guessing to inspect the `struct file` received as a parameter of `fd_install`.

**Comment by alban** [comment]
I rebased this on top of https://github.com/weaveworks/tcptracer-bpf/pull/37 because I needed to revendor gobpf for https://github.com/iovisor/gobpf/pull/45 but this also brings other gobpf API changes that cause tcptracer-bpf changes that were already done in #37.

File: tests/tracer.go [hunk 1]
```diff
@@ -41,8 +42,8 @@ func lostCb(count uint64) {
 }
 
 func main() {
-	if len(os.Args) != 1 {
-		fmt.Fprintf(os.Stderr, "Usage: %s\n", os.Args[0])
+	if len(os.Args) > 2 {
+		fmt.Fprintf(os.Stderr, "Usage: %s pid\n", os.Args[0])
```
Comment (schu) [inline tests/tracer.go]: Nit: should be `[pid]`

Comment (alban) [reply tests/tracer.go]: I refactored the code using Golang's flag so I don't have to write the "usage" myself.

**Comment by alban** [comment]
I added a test and it passes in Semaphore with Linux 4.4.45 and 4.9.6. It works as well on my laptop with Linux 4.10.10.

**Comment by alban** [comment]
I updated this PR:
- fix compilation on Darwin (revealed by Scope tests in CircleCI)
- added more shellcheck fixes (it will need a proper "git subtree" update after https://github.com/weaveworks/build-tools/pull/97 is merged)

**Comment by alban** [comment]
Since weaveworks/build-tools#97 is merged, I rebased & updated the "git subtree". But this time, there are new errors (see https://github.com/weaveworks/build-tools/pull/99)

**Comment by alban** [comment]
I added a commit that fixes the error. The Dockerfile now uses shfmt with another importpath. Thanks @tomwilkie for the hint in https://github.com/weaveworks/build-tools/pull/99#issuecomment-302365127

@2opremio PTAL

File: pkg/tracer/event_common.go [hunk 2]
```diff
@@ -8,9 +8,10 @@ type EventType uint32
 
 // These constants should be in sync with the equivalent definitions in the ebpf program.
 const (
-	EventConnect EventType = 1
-	EventAccept            = 2
-	EventClose             = 3
+	EventConnect   EventType = 1
```
Comment (2opremio) [inline pkg/tracer/event_common.go:11]: I didn't even know about this syntax

File: tcptracer-bpf.c [hunk 3]
```diff
@@ -821,6 +841,48 @@ int kretprobe__inet_csk_accept(struct pt_regs *ctx)
 	return 0;
 }
 
+SEC("kprobe/fd_install")
+int kprobe__fd_install(struct pt_regs *ctx)
+{
+	u64 pid = bpf_get_current_pid_tgid();
+	u32 tgid = pid >> 32;
+	unsigned long fd = (unsigned long) PT_REGS_PARM1(ctx);
```
Comment (2opremio) [inline tcptracer-bpf.c:849]: I thought LLVM automatically did the parameter mapping (e.g. [like this](https://github.com/weaveworks-plugins/scope-http-statistics/blob/master/ebpf-http-statistics.c#L25)).

Is there a reason for using the macro instead? (using parameters is more readable)

Seeing the rest of the kprobes, I realize I am coming late with this remark, sorry.

Comment (alban) [reply tcptracer-bpf.c:849]: I think this parameter mapping is done by the bcc framework.

**Comment by 2opremio** [comment]
Minor comment, otherwise LGTM

File: tcptracer-bpf.c (hunk 3 above)
Comment (2opremio) [reply tcptracer-bpf.c:849]: I would be surprised if this didn't happen through LLVM, but you know it better :)

Comment (alban) [reply tcptracer-bpf.c:849]: It might be LLVM code in bcc that does the rewriting, but I am not sure where exactly.

**Comment by alban** [comment]
@2opremio I think I replied to the comment. Someone with merge-powers could merge this :)

File: tcptracer-bpf.c (hunk 3 above)
Comment (iaguis) [reply tcptracer-bpf.c:849]: Probably [here](https://github.com/iovisor/bcc/blob/381f03dbdb2d5723cb585c241bf05c6fcad8dd7a/src/cc/frontends/clang/loader.cc) and [here](https://github.com/iovisor/bcc/blob/381f03dbdb2d5723cb585c241bf05c6fcad8dd7a/src/cc/frontends/clang/b_frontend_action.cc#L281)

Comment (2opremio) [reply tcptracer-bpf.c:849]: I know we are not using bcc but I was assuming we were using the same C compiler. We don't necessarily need to change that, but I would like to understand the differences. Mind elaborating?

**Comment by 2opremio** [comment]
Merging, but let's continue the comment thread on LLVM please.

**Review by 2opremio — APPROVED** [review]

File: tcptracer-bpf.c (hunk 3 above)
Comment (iaguis) [reply tcptracer-bpf.c:849]: It is the same compiler but bcc implements a (actually several) rewrite passes on the source files by using clang's API. See http://eli.thegreenplace.net/2012/06/08/basic-source-to-source-transformation-with-clang for example.

Comment (2opremio) [reply tcptracer-bpf.c:849]: Aha, and I guess there is no gain in using those here, right?

Comment (iaguis) [reply tcptracer-bpf.c:849]: I think the complexity and the time we'll spend extracting that from bcc and applying it here is not worth it.

In the end, the rewriting is done to make creating bpf programs easier and reduce verifier errors, but we already did the effort of writing pure bpf and making it compliant.

Comment (2opremio) [reply tcptracer-bpf.c:849]: Fair enough. Thanks a lot for the clarifications.

**Pull request merged**